File: containsDuplicate.py
#Hashmap/Dictionary Solution:
#Time complexity = O(n) because we iterate over the list once and perform O(1) operations inside the loop. Hence O(1) * O(n) = O(n)
#Space complexity = O(n) because for worse case, we store all n unique numbers in the dictionary

from typing import List #import this module and use this specific class
class Solution:
    def hasDuplicate(self, nums: List[int]) -> bool:
#optimized solution in O(n) time and 0(n) space complexity
        dictionary = {}
        for (number : nums):
            if(number not in dictionary):
                dictionary[number] = 1
            else:
                return True
        return 
        


##########################################################
##########################################################

#Hashset Solution:
#Time Complexity -> O(n)
#Space Complexity -> O(n)
from typing import List
class Solution:
    def hasDuplicate(self, List[int] nums):
        checkedAlready = set()
        for(number : nums):
            if(number in checkedAlready):
                return True
            else:
                checkedAlready.add(number)
        return False

##########################################################
##########################################################

#Brute force
#Time complexity -> O(n^2)
#Space complexity -> O(1)
from typing import List
class Solution:
    def hasDuplicate(self, List[int] nums):
        for i in range(0, len(nums), 1):
            for j in range(i+1, len(nums), 1):
                if nums[i] == nums[j]:
                    return True
        return False

File: containsDuplicate.cpp
//Using hashset
//Time Complexity -> O(n) because we are using a for each loop
//Space Complexity -> O(n) in memory bc we are storing values in an unordered set
#include <vector>
using namespace std; // not use std::
//We don't use an unordered_map because it stores key-value pairs, but we can use unordered_set to check for the unique key's existence.
class Solution {
    bool hasDuplicate(vector<int>& nums){
        unordered_set<int> checkedAlready; //would've been std::unordered_set<int> but is not because of using namespace std;
        for(num : nums){
            if(checkedAlready.count(num)) {
                return true;
            }
            checkAlready.insert(num);
        }
        return false;
    }
}

/////////////////////////////////////////////////////
/////////////////////////////////////////////////////

//Brute force
//Time Comp. -> O(n^2) bc uses 2 for loops nested
//Space Comp. -> O(1) bc it doesn't initialize any arr, and only returns true/false
#include <vector>
class Solution {
public:
    bool hasDuplicate(vector<int>& nums) {
        for(int i =0; i < nums.size(); i++){
            for(int j = i + 1; j < nums.size(); j++){
                if (nums[i] == nums[j]){
                    return true;
                }
            }
        }
        return false;
    }
};

/////////////////////////////////////////////////////
/////////////////////////////////////////////////////

//Use Sorting Algorithm
//Time Complexity -> O(nlogn) bc we are using a sorting alorithm
//Space Complexity -> O(1) in memory bc we aren't using any other data structures
#include <vector>
#include <algorithm>

class Solution {
public:
    bool hasDuplicate(vector<int>& nums){
        std::sort(nums.begin(), nums.end());
        for(int i = 0; i < nums.size()-1; i++) {
            if (nums[i] == nums[i+1]) {
                return true;
            }
        }
        return false;
    }
};